feat(code): Vote extensions v2

romac opened the pull request:
Closes: #717 

---

## Questions

- [x] Should the app generate a plain `Extension` and then have it signed by consensus or generate a `SignedExtension` directly?
  - Let's have consensus sign and verify vote extensions

---

### PR author checklist

#### For all contributors

- [x] Reference GitHub issue
- [x] Ensure PR title follows the [conventional commits][conv-commits] spec

#### For external contributors

- [ ] Maintainers of Malachite are [allowed to push changes to the branch][gh-edit-branch]

[conv-commits]: https://www.conventionalcommits.org/en/v1.0.0/#summary
[gh-edit-branch]: https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests

ancazamfir commented:
Looks good so far. 
Previous implementation was not really a vote extension ala Comet. And we were not doing anything at all with the extensions. We built that in order to compare the case where all Tx-es are included in the block by the proposer and therefore disseminated by a single sender (the proposer) with the case where the Tx-es (for next block) are included in the votes, i.e. Tx dissemination by multiple senders (potentially all validators). For the latter case, mempool bcast was disabled. We measured and compared the latencies but those measurements did not include Tx execution or block store times (we didn't have it at the time).
Given this, in v2 we need to pass the extensions to the app (starknet) that will do something with them. Here are a few comments along those lines:

- In the starknet app we can have a usecase along the lines of the e2e app in Comet, see:
https://github.com/cometbft/cometbft/blob/63a2a645835fcba5ac7ff6926116d49952e49ca4/test/e2e/app/app.go#L608-L627

- [on_decided()](https://github.com/informalsystems/malachite/blob/f938db651df80177031287fc05f03e5877f1ffc9/code/crates/starknet/host/src/actor.rs#L548)  - Should take an extra argument, the extensions seen by this node. The vote signature should not include the extension (keep the extension signature separate), the certificate should only include/ aggregate the votes. 
The extensions should be stored but not part of the block.

- If we are proposer in the next block, in [on_get_value()](https://github.com/informalsystems/malachite/blob/f938db651df80177031287fc05f03e5877f1ffc9/code/crates/starknet/host/src/actor.rs#L277) we build a transaction with those extensions and include it after the other tx-es. Alternatively we pass the extensions to `on_get_value()`. In this case we have to keep the extensions in consensus (the driver) when moving to the next height.

ancazamfir commented:
regarding @cason [comment](https://github.com/informalsystems/malachite/issues/717#issuecomment-2572981285)
Maybe we can include that here as well. This way we can clean the `(Signed)Vote` protos and code and add a separate message `(Signed)VoteExtension`. We can diseminate over a different topic for now.

romac commented:
> regarding @cason [comment](https://github.com/informalsystems/malachite/issues/717#issuecomment-2572981285) Maybe we can include that here as well. This way we can clean the `(Signed)Vote` protos and code and add a separate message `(Signed)VoteExtension`. We can diseminate over a different topic for now.

I agree we should do that, but only in a follow-up PR.

ancazamfir commented:
> > regarding @cason [comment](https://github.com/informalsystems/malachite/issues/717#issuecomment-2572981285) Maybe we can include that here as well. This way we can clean the `(Signed)Vote` protos and code and add a separate message `(Signed)VoteExtension`. We can diseminate over a different topic for now.
> 
> I agree we should do that, but only in a follow-up PR.

That also works so let's do that if you think it's better. Was about to say the same but then I thought that a lot of vote structure and handling will need to be undone.

romac commented:
> `on_decided` [...] should take an extra argument, the extensions seen by this node

Aren't the extensions already included in the certificate passed to `on_decided`? ie. `certificate.aggregated_signature.signatures[].extension`

romac commented:
> Was about to say the same but then I thought that a lot of vote structure and handling will need to be undone.

Yes perhaps but since doing separate propagation of vote extensions might end up being complex I'd be more comfortable doing it in its own PR even if I have to undo a bunch of changes made here.

romac commented:
> * The extensions should be stored but not part of the block.

You mean not store in the db in the `decided_blocks` table but rather in a separate table, or just kept in memory?

ancazamfir commented:
> > * The extensions should be stored but not part of the block.
> 
> You mean not stored in the db in the `decided_blocks` table but rather in a separate table, or just kept in memory?

just in memory is fine, given how we use them in the mock app, i don't think there is a problem if we lose them on restart.

romac commented:
@ancazamfir Let me know if my latest commits have addressed your comments. If so then, I'll get to including the extensions in the value itself (maybe as a special tx for the Starknet app)

ancazamfir commented:
> @ancazamfir Let me know if my latest commits have addressed your comments. If so then, I'll get to including the extensions in the value itself (maybe as a special tx for the Starknet app)

yes, looks good!

ancazamfir commented on code/crates/starknet/host/src/host/proposal.rs:
```diff
@@ -149,6 +152,22 @@ async fn run_build_proposal_task(
+            .map(Transaction::new)
+            .collect();
+
+        let part = ProposalPart::Transactions(Transactions::new(transactions));
```
maybe make sure they're in multiple parts if we're over the part size.

romac replied:
Good point! Moreover, what happens if `value_size + extensions_size > max_block_size`, do we just drop the remaining extensions or do we include them regardless? cc @cason

ancazamfir replied:
let's just drop. not much else we can do. Note this is just something for our test app so we can do whatever we want here. In a useful app things would be different but wanted something very simple.

cason replied:
Why do we broadcast the extensions in the proposal? It should be enough to include the hashes of them. Nodes are supposed to receive them independently.

romac replied:
@ancazamfir I just realized we do not really have a `max_part_size`, but rather `max_txs_per_part`. So we can either:

- Use some heuristic to compute the `max_part_size` based on txs seen so far
- Add a `max_part_size` setting, remove `max_txs_per_part`
- Have one part per extension
- Leave it as is

What do you think?

romac replied:
> Why do we broadcast the extensions in the proposal? It should be enough to include the hashes of them. Nodes are supposed to receive them independently.

Yes, I'll get to that in a separate PR. For now we are just piggybacking on proposal parts gossip to disseminate the extensions.

ancazamfir replied:
> @ancazamfir I just realized we do not really have a `max_part_size`, but rather `max_txs_per_part`. So we can either:
> 
> * Use some heuristic to compute the `max_part_size` based on txs seen so far
> * Add a `max_part_size` setting, remove `max_txs_per_part`
> * Have one part per extension
> * Leave it as is
> 
> What do you think?

Do we have access to the `vote_extensions.size` config? If not we can just count one extension as one tx for now. And maybe in the app when we generate it we use min(`tx_size`, `vote_extensions.size`) since the extension is used as a tx in this app.

cason commented:
Jumping a little late here, vote extensions are transitional data within a height. They should not be stored on chain, they should not, except under request, be exchanged by peers (via VoteSync). They don't have any use from when the next proposal is produced and/or received.

The idea of having two signatures, one for the vote extension, another for the vote message itself is not great, to be honest. In my opinion, extensions are data produced by the application, that might choose or not to sign it. The vote message could include the hash of the vote extension, a constant-size value, with a special value for no/empty extension. This saves us from signing votes twice _but_ is a problem if we want to consider vote/signatures aggregation, since votes for the same round step and value become different, since the extensions are non-deterministic data. This topic is worth additional discussion. The solution adopted by Comet is sub-optimal, in my view, it is like that for "historical reasons".

cason commented:
> [on_decided()](https://github.com/informalsystems/malachite/blob/f938db651df80177031287fc05f03e5877f1ffc9/code/crates/starknet/host/src/actor.rs#L548) - Should take an extra argument, the extensions seen by this node.

Notice that in this case, each node is seeing "only" strictly 2/3+ of the produced vote extensions. The solution works, in terms of performance improvements, mostly if we can see more than that, ideally all produced vote extensions.

Also, from the design concept, they are really only useful for the next height, either when producing a block, or when validating a block produced by another node.

romac commented:
> Notice that in this case, each node is seeing "only" strictly 2/3+ of the produced vote extensions

How about when running with non-zero commit timeout? Shouldn't this let us gather more precommits and therefore more vote extensions?

romac commented:
> Jumping a little late here, vote extensions are transitional data within a height. They should not be stored on chain, they should not, except under request, be exchanged by peers (via VoteSync). They don't have any use from when the next proposal is produced and/or received.

Yep that's the way they are now implemented as of this PR.

> The idea of having two signatures, one for the vote extension, another for the vote message itself is not great, to be honest. In my opinion, extensions are data produced by the application, that might choose or not to sign it. 

That's a valid point. So you are saying we should only work with the abstract `Extension` type and leave it up to the application to instantiate it with a signed or unsigned extension type?

> The vote message could include the hash of the vote extension, a constant-size value, with a special value for no/empty extension. 

This will only apply when we start disseminating vote extensions over a bespoke channel instead of stuffing them in the votes, right?

I was actually wondering about that, thinking that we need a way to know how to correlate an extension we receive "out-of-band" (ie. not via the consensus gossip topic but over some other topic/channel, eg. DOG) with a precommit. Including the hash of the vote extension in the vote seems like a good idea and would allow us to track whether we have received all extensions for votes we have seen and to discard extensions for votes we have not seen.
Does that makes sense?

cason commented:
> How about when running with non-zero commit timeout? Shouldn't this let us gather more precommits and therefore more vote extensions?

Yes, but we should not wait for the commit timeout to decide, but to start the next height.

cason commented:
> > How about when running with non-zero commit timeout? Shouldn't this let us gather more precommits and therefore more vote extensions?
> 
> Yes, but we should not wait for the commit timeout to decide, but to start the next height.

And in anyway this is not a really nice design. There are other ways to collect additional precommits without halting the node for 1s.

cason commented:
> > The idea of having two signatures, one for the vote extension, another for the vote message itself is not great, to be honest. In my opinion, extensions are data produced by the application, that might choose or not to sign it.
> 
> That's a valid point. So you are saying we should only work with the abstract `Extension` type and leave it up to the application to instantiate it with a signed or unsigned extension type?
> 
> > The vote message could include the hash of the vote extension, a constant-size value, with a special value for no/empty extension.
> 
> This will only apply when we start disseminating vote extensions over a bespoke channel instead of stuffing them in the votes, right?
> 
> I was actually wondering about that, thinking that we need a way to know how to correlate an extension we receive "out-of-band" (ie. not via the consensus gossip topic but over some other topic/channel, eg. DOG) with a precommit. Including the hash of the vote extension in the vote seems like a good idea and would allow us to track whether we have received all extensions for votes we have seen and to discard extensions for votes we have not seen. Does that makes sense?

Yes, this is exactly the point. We can propagate extensions as we want, the most efficient method for it, but we should also be able to link an extension to a vote. This mean that when we receive a vote (not necessarily a precommit, could be also a Prevote) we need to know that there is an extension `X` we should expect from the same sender.

I really don't see how to "ensure" that without having something like `hash(X)` as part of the vote. Which has, on the other hand, the pitfall of rendering vote/signature aggregation more complex.

ancazamfir commented:
> Which has, on the other hand, the pitfall of rendering vote/signature aggregation more complex.

I agree with this but not clear to me we need a hash, depends on the requirements. 

When an extension comes, since it would include (height, round, id), we can drop it if it's invalid (e.g. we moved to the next height or not coming from a val in the set). At "decided" time we can exclude the ones that don't match the commit certificate (height, round, id) or if the val has not voted (the extension may arrive before the vote)

Should it be mandatory for a validator to issue an extension (if enabled)? Could we just collect extensions on a best effort basis, meaning there can be less than 2/3rds?

I propose we write a spec for VE v3 and merge this PR for now.

cason commented:
> I propose we write a spec for VE v3 and merge this PR for now.

I agree.

cason commented:
> Should it be mandatory for a validator to issue an extension (if enabled)? Could we just collect extensions on a best effort basis, meaning there can be less than 2/3rds?

We can do in a best-effort in a first effort. But recall that a Byzantine validator can produce multiple extensions for the same round step (height, round, step). And if they are not signed, it can even produce extensions on behalf of other validators...

Plus, if this is a best-effort dissemination of data, what is the actual difference from a mempool? The fact that we expect at most one message per sender?

ancazamfir commented:
> > Should it be mandatory for a validator to issue an extension (if enabled)? Could we just collect extensions on a best effort basis, meaning there can be less than 2/3rds?
> 
> We can do in a best-effort in a first effort. But recall that a Byzantine validator can produce multiple extensions for the same round step (height, round, step). 

We can just keep one if multiple are sent. 

> And if they are not signed, it can even produce extensions on behalf of other validators...
> 

They should be signed.

> Plus, if this is a best-effort dissemination of data, what is the actual difference from a mempool? The fact that we expect at most one message per sender?

Yes. And if an app wants to collect more it can set a bigger commit timeout. 

But let's write a spec and if possible not include anything in the vote. My concern is that our main customer is starknet and they haven't express any interest in vote extension so ideally we should use their proto definitions.

ancazamfir reviewed (approved):
looks great! 🚀
